src/datapoints-graph/charts/echarts-options.service.ts

```diff
@@ -111,24 +111,242 @@ export class EchartsOptionsService {
         },
       },
       yAxis,
-      series: this.getChartSeries(datapointsWithValues),
+      series: this.getChartSeries(datapointsWithValues, events, alarms),
     };
   }
 
   private getChartSeries(
-    datapointsWithValues: DatapointWithValues[]
+    datapointsWithValues: DatapointWithValues[],
+    events,
+    alarms
   ): SeriesOption[] {
     const series: SeriesOption[] = [];
+    let eventSeries: SeriesOption[] = [];
+    let alarmSeries: SeriesOption[] = [];
     datapointsWithValues.forEach((dp, idx) => {
       const renderType: DatapointChartRenderType = dp.renderType || 'min';
       if (renderType === 'area') {
         series.push(this.getSingleSeries(dp, 'min', idx, true));
         series.push(this.getSingleSeries(dp, 'max', idx, true));
       } else {
-        series.push(this.getSingleSeries(dp, renderType, idx));
+        series.push(this.getSingleSeries(dp, renderType, idx, false));
       }
+
+      const newEventSeries = this.getAlarmOrEventSeries(
+        dp,
+        renderType,
+        false,
+        events,
+        'event'
+      );
+      const newAlarmSeries = this.getAlarmOrEventSeries(
+        dp,
+        renderType,
+        false,
+        alarms,
+        'alarm'
+      );
+      eventSeries = [...eventSeries, ...newEventSeries];
+      alarmSeries = [...alarmSeries, ...newAlarmSeries];
     });
-    return series;
+    return [...series, ...eventSeries, ...alarmSeries];
+  }
+
+  groupByType(items: any[], typeField: string, hiddenField: string) {
+    return items.reduce((grouped, item) => {
+      if (!item[hiddenField]) {
+        (grouped[item[typeField]] = grouped[item[typeField]] || []).push(item);
+      }
+      return grouped;
+    }, {});
+  }
+
+  getClosestDpValue(dpValuesArray: any, targetTime: number) {
+    return dpValuesArray.reduce((prev, curr) =>
+      Math.abs(curr.time - targetTime) < Math.abs(prev.time - targetTime)
+        ? curr
+        : prev
+    );
+  }
+
+  createMarkPoint(item: any, dp: any, isCleared: boolean, isEvent: boolean) {
```

Comment:
we can definitely can tell these types and not use 'any'. also add return type and some ts-docs describing what is mark point